job-server/src/main/scala/spark/jobserver/JobManager.scala

```diff
@@ -70,28 +58,37 @@ object JobManager {
 
     logger.info("Starting JobManager named " + managerName + " with config {}",
       config.getConfig("spark").root.render())
-    logger.info("..and context config:\n" + contextConfig.root.render)
 
-    val jobManager = system.actorOf(JobManagerActor.props(contextConfig, daoActor), managerName)
+    val jobManager = system.actorOf(JobManagerActor.props(daoActor), managerName)
 
     //Join akka cluster
     logger.info("Joining cluster at address {}", clusterAddress)
     Cluster(system).join(clusterAddress)
 
-    //Kill process on actor system shutdown
     val reaper = system.actorOf(Props[ProductionReaper])
-    system.registerOnTermination(System.exit(0))
     reaper ! WatchMe(jobManager)
+
+    waitForTermination(system, deployMode)
   }
 
   def main(args: Array[String]) {
     import scala.collection.JavaConverters._
+
     def makeManagerSystem(name: String)(config: Config): ActorSystem = {
       val configWithRole = config.withValue("akka.cluster.roles",
         ConfigValueFactory.fromIterable(List("manager").asJava))
       ActorSystem(name, configWithRole)
     }
-    start(args, makeManagerSystem("JobServer")(_))
+
+    def waitForTermination(system: ActorSystem, deployMode: String) {
+      if (deployMode == "cluster") {
+        system.awaitTermination // Wait for actor system shutdown
```

Comment:
This line is causing the problem.
If I replace this line with `system.registerOnTermination(System.exit(0))` then the problem is resolved for cluster mode.

It seems that the same problem should also exist for YARN mode. Could you check ? My steps
- Start a streaming context
- Start a job
- Delete the context
- Check master UI and the driver process is still there

Now, why `system.awaitTermination` is not working? I am not fully sure about the lifecycle after cluster.down() is called from AkkaClusterSupervisorActor. It seems that some operations are still being done using SparkContext but since normally we execute `System.exit(0)`, we don't see the error.

Reply:
@velvia Could you think of a reason why using `system.awaitTermination` leads to exception posted in my previous comment ?

Reply:
@bsikander Thanks for pointing this out. I tested the YARN cluster mode with a batch job and the old behavior (`system.registerOnTermination(System.exit(0))`) results in a immediately finished context. It looks like that the finished main method also finished the YARN job. The `system.awaitTermination` avoids this and keeps the method running until the context has finished. Sound like we have to change this to something like `deployMode == "cluster" && master == "yarn"` right?

Reply:
I do not have much experience in YARN so cannot say but if this check fixes the problem then we should do it.

+1 for the update.